src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php

```diff
@@ -0,0 +1,129 @@
+    }
+
+    /**
+     * Finds the correct authenticator for the token and calls it.
+     *
+     * @param GuardTokenInterface $token
+     *
+     * @return TokenInterface
+     */
+    public function authenticate(TokenInterface $token)
+    {
+        if (!$this->supports($token)) {
+            throw new \InvalidArgumentException('GuardAuthenticationProvider only supports NonAuthenticatedGuardToken');
+        }
+
+        if (!$token instanceof PreAuthenticationGuardToken) {
+            /*
+             * The listener *only* passes PreAuthenticationGuardToken instances.
+             * This means that an authenticated token (e.g. PostAuthenticationGuardToken)
+             * is being passed here, which happens if that token becomes
+             * "not authenticated" (e.g. happens if the user changes between
+             * requests). In this case, the user should be logged out, so
+             * we will return an AnonymousToken to accomplish that.
+             */
+
+            // this should never happen - but technically, the token is
+            // authenticated... so it could just be returned
+            if ($token->isAuthenticated()) {
+                return $token;
+            }
+
+            // cause the logout - the token is not authenticated
+            return new AnonymousToken($this->providerKey, 'anon.');
+        }
+
+        // find the *one* GuardAuthenticator that this token originated from
+        foreach ($this->guardAuthenticators as $key => $guardAuthenticator) {
+            // get a key that's unique to *this* guard authenticator
+            // this MUST be the same as GuardAuthenticationListener
+            $uniqueGuardKey = $this->providerKey.'_'.$key;
+
+            if ($uniqueGuardKey == $token->getGuardProviderKey()) {
+                return $this->authenticateViaGuard($guardAuthenticator, $token);
+            }
+        }
+
+        throw new \LogicException(sprintf(
+            'The correct GuardAuthenticator could not be found for unique key "%s". The listener and provider should be passed the same list of authenticators!?',
+            $token->getGuardProviderKey()
+        ));
+    }
+
+    private function authenticateViaGuard(GuardAuthenticatorInterface $guardAuthenticator, PreAuthenticationGuardToken $token)
+    {
+        // get the user from the GuardAuthenticator
+        $user = $guardAuthenticator->authenticate($token->getCredentials(), $this->userProvider);
+
+        if (!$user instanceof UserInterface) {
+            throw new \UnexpectedValueException(sprintf(
+                'The %s::authenticate method must return a UserInterface. You returned %s',
+                get_class($guardAuthenticator),
+                is_object($user) ? get_class($user) : gettype($user)
+            ));
+        }
+
+        // check the AdvancedUserInterface methods!
+        $this->userChecker->checkPreAuth($user);
+        $this->userChecker->checkPostAuth($user);
```

Comment:
A lot here :)

1) Actually, `checkPostAuth()` is right, but `checkPreAuth()` was wrong - it should happen right after fetching the User, but before checking credentials (see `UserAuthenticationProvider`). I've just split a method on the interface to make this possible and moved the `checkPreAuth()` in the provider. Thanks for asking about this - I hadn't really realized I had things in the wrong spot.

2) About the IP timeout idea, was this simpler with the simple-authenticator? Or are you saying that in both cases, you'd need to use inheritance or duplication to put the code into the "simple authenticator" or the "guard authenticator", so it's the same, but no better? I just commented on #11090 - I think it can be solved there.

3) We should avoid using forms or validation anywhere. But more importantly, the "guard authenticators" are so simple that you can do whatever you want, including using forms and validation.